Implement `Random` for array

sorairolake opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/136732)*
<!-- homu-ignore:end -->

Implement `Random` for `[T; N]`, where `Random` is implemented for `T`.

<https://github.com/rust-lang/libs-team/issues/393#issue-2345931645> states:

> The trait `Random` will initially be implemented for all iN and uN integer types, `isize`/`usize`, and `bool`, as well as **arrays**** and tuples of such values.

This PR is also based on <https://github.com/rust-lang/rust/issues/130703#issuecomment-2639855152>.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Tracking issue: rust-lang/rust#130703

sorairolake committed:
Implement `Random` for array

Implement `Random` for `[T; N]`, where `Random` is implemented for `T`.

sorairolake commented:
r? @joboet

joboet commented:
I think it would be better to implement this with a dedicated method on the `Random` trait, or at least adding a specialization for the integer primitives. Otherwise generating e.g. a `[u8; _]` will call into the system for each element instead of just once to fill the entire buffer, which is very wasteful and slow.

Urgau commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,18 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        let mut buf = [const { MaybeUninit::uninit() }; N];
+        for elem in &mut buf {
+            elem.write(T::random(source));
```
If `T::random` panics the elements already present in `buf` won't be freed.

We generally try to avoid leaking memory when possible, we generally do that with a guard type that frees on `Drop`, and is `mem::forgot` when everything is fine.

We already have such a [`Guard` type](https://github.com/rust-lang/rust/blob/d2f335d58e6c346f94910d0f49baf185028b44be/library/core/src/array/mod.rs#L870-L886) for uninit-arrays, see [here](https://github.com/rust-lang/rust/blob/d2f335d58e6c346f94910d0f49baf185028b44be/library/core/src/array/mod.rs#L857-L866) for how to use it.

joboet commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,18 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+#[unstable(feature = "random", issue = "130703")]
+impl<T: Random, const N: usize> Random for [T; N] {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        let mut buf = [const { MaybeUninit::uninit() }; N];
+        for elem in &mut buf {
+            elem.write(T::random(source));
+        }
+        // SAFETY: all elements of the array were initialized.
+        unsafe { mem::transmute_copy(&buf) }
```
Why not just use `array::from_fn`:
```suggestion
        array::from_fn(|_| T::random(source))
```

sorairolake replied:
I didn't know about this convenient function.

the8472 commented:
>  Otherwise generating e.g. a [u8; _] will call into the system for each element instead of just once to fill the entire buffer, which is very wasteful and slow.

That doesn't necessarily need specialization. We could add an internal `BufferedRandomSource` akin to a `BufReader` and then discard it at the end of the array filling.
We only want the OS source to be ~unbuffered to avoid fork/VM cloning issues. But if you're cloning a VM in the middle of array-filling you'll already end up with some duplicated "random" data anyway, so this should be fine I think.

abgros commented:
Calling `random()` for every element in the array is extremely wasteful because as @joboet said it does a separate system call for each element. I benchmarked the following functions using criterion:
```rs
fn random_array() -> [u8; 1_000] {
	array::from_fn(|_| u8::random(&mut DefaultRandomSource))
}

fn random_array2() -> [u8; 1_000] {
	let mut arr = [0; 1_000];
	DefaultRandomSource.fill_bytes(&mut arr);
	arr
}
```
Calling `fill_bytes()` a single time is over 50x faster.
```
random_array            time:   [42.631 µs 43.067 µs 43.875 µs]
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

random_array2           time:   [617.27 ns 626.50 ns 641.50 ns]
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe
```

abgros commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,18 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+#[unstable(feature = "random", issue = "130703")]
+impl<T: Random, const N: usize> Random for [T; N] {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        let mut buf = [const { MaybeUninit::uninit() }; N];
+        for elem in &mut buf {
+            elem.write(T::random(source));
+        }
+        // SAFETY: all elements of the array were initialized.
+        unsafe { mem::transmute_copy(&buf) }
```
Here's a version that uses `fill_bytes()`.
```suggestion
        let mut buf = [const { MaybeUninit::<T>::uninit() }; N];
        // SAFETY: this initializes every element in the buffer with random bytes.
        unsafe {
            let slice = core::slice::from_raw_parts_mut(&raw mut buf as *mut u8, N * mem::size_of::<T>());
            source.fill_bytes(slice);
            mem::transmute_copy(&buf)
        }
```

the8472 replied:
This is unsound for generic Ts. There's no guarantee that a Type's `random` impl shovels those bytes into their memory representation. A trivial example would be `NonZero<u8>` which could implement `Random` via rejection sampling. Setting it to 0 would be unsound.

abgros replied:
@the8472 I did notice that, so we would need to restrict this to types that can hold arbitrary bytes (which currently includes every type implementing `Random` except `bool`). Otherwise, use a slower custom implementation with rejection sampling or modulo.

sorairolake committed:
Change to use `array::from_fn`

sorairolake committed:
Add specialization `Random` for integer arrays

Implements specialized `Random` for the array whose elements are
integer primitives. This reduces the number of system calls.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13221271768/job/36906612317) [(plain)](https://github.com/rust-lang/rust/commit/7c195abdeb47d6734cc3dff2cace946b8e439e05/checks/36906612317/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling object v0.36.7
error: found a count of bytes instead of a count of elements of `T`
   --> library/core/src/array/mod.rs:448:25
    |
437 | / macro_rules! impl_random_for_integer_array {
438 | |     ($t:ty) => {
439 | |         #[unstable(feature = "random", issue = "130703")]
440 | |         impl<const N: usize> Random for [$t; N] {
...   |
448 | |                         N * mem::size_of::<$t>(),
...   |
455 | |     };
456 | | }
    | |_- in this expansion of `impl_random_for_integer_array!`
    | |_- in this expansion of `impl_random_for_integer_array!`
457 |
458 |   impl_random_for_integer_array!(u8);
    |
    = help: use a count of elements instead of a count of bytes, it already gets multiplied by the size of the type
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#size_of_in_element_count
    = note: `-D clippy::size-of-in-element-count` implied by `-D clippy::correctness`
```

</details>

the8472 commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,47 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+macro_rules! impl_random_for_integer_array {
+    ($t:ty) => {
+        #[unstable(feature = "random", issue = "130703")]
+        impl<const N: usize> Random for [$t; N] {
```
Specializations should be hidden behind a private trait. https://std-dev-guide.rust-lang.org/policy/specialization.html

Alternatively you can avoid specialization by implementing a buffering random source wrapper as mentioned in https://github.com/rust-lang/rust/pull/136732#issuecomment-2645871365

sorairolake committed:
Hide specialized `Random` for array behind private trait

joboet commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,59 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+                        &raw mut buf as *mut u8,
+                        N * (<$t>::BITS as usize / 8),
+                    );
+                    source.fill_bytes(bytes);
```
This is unsound: you create a `&mut [u8]` and pass it to user code – but nothing guarantees that the user code won't try to read the slice, which would be undefined behaviour. My recommendation would be to use `MaybeUninit::zeroed` to create `buf`.

dhardy replied:
Ideally this needs a trait bound like [`zerocopy::FromBytes`](https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html) or [`bytemuck::Pod`](https://docs.rs/bytemuck/latest/bytemuck/trait.Pod.html), but we don't have that in `std`, so... the macro should be renamed to `unsafe_impl_...` since safety depends on the macro caller?

joboet commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,59 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+                // SAFETY: all elements in the buffer were initialized with
+                // random bytes.
+                unsafe {
+                    let bytes = slice::from_raw_parts_mut(
+                        &raw mut buf as *mut u8,
+                        N * (<$t>::BITS as usize / 8),
+                    );
```
```suggestion
                    let bytes = buf.as_bytes_but().assume_init_mut();
```

joboet commented on library/core/src/array/mod.rs:
```diff
@@ -426,6 +427,59 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
+                        N * (<$t>::BITS as usize / 8),
+                    );
+                    source.fill_bytes(bytes);
+                    mem::transmute_copy(&buf)
```
```suggestion
                    MaybeUninit::array_assume_init(buf)
```

joboet reviewed (changes requested):
I too would prefer a buffering implementation (in the generic case, the specialization for primitives makes sense to me) – but you can leave that for another PR.

sorairolake committed:
Apply suggestions from code review

dhardy commented:
This uses specialization to do what we couldn't do with [`rand::Fill`](https://docs.rs/rand/latest/rand/trait.Fill.html), which is great except:

Any non-default impls of `Random` will produce different values when sampled. This makes the introduction of new non-default impls potentially a breaking change, depending on how `std` wants to handle reproducibility guarantees for random generators.

(My view is still that random-value generation should be left to external crates like `rand`, partly to avoid this reproducibility-guarantee issue.)

bors commented:
:umbrella: The latest upstream changes (presumably #135634) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Amanieu commented:
Note that we are close to landing support for arbitrary types in `BufReader` in #149749, which would allow implementing the proposal in https://github.com/rust-lang/rust/pull/136732#issuecomment-3378060391.
